Yosemite/Yosemite/Stores/SystemPluginStore.swift

```diff
@@ -0,0 +1,102 @@
+            fetchSystemPlugins(siteID: siteID, completionHandler: onCompletion)
+        }
+    }
+}
+
+// MARK: - Network request
+//
+private extension SystemPluginStore {
+    func synchronizeSystemPlugins(siteID: Int64, completionHandler: @escaping (Result<Void, Error>) -> Void) {
+        remote.loadSystemPlugins(for: siteID) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let systemPlugins):
+                self.upsertSystemPluginsInBackground(siteID: siteID, readonlySystemPlugins: systemPlugins, completionHandler: completionHandler)
+            case .failure(let error):
+                completionHandler(.failure(error))
+            }
+        }
+    }
+}
+
+// MARK: - Storage
+//
+private extension SystemPluginStore {
+
+    /// Updates or inserts Readonly `SystemPlugin` entities in background.
+    /// Triggers `completionHandler` on main thread.
+    ///
+    func upsertSystemPluginsInBackground(siteID: Int64, readonlySystemPlugins: [SystemPlugin], completionHandler: @escaping (Result<Void, Error>) -> Void) {
+        let writerStorage = storageManager.writerDerivedStorage
+        writerStorage.perform {
+            self.upsertSystemPlugins(siteID: siteID, readonlySystemPlugins: readonlySystemPlugins, in: writerStorage)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: writerStorage) {
+            DispatchQueue.main.async {
+                completionHandler(.success(()))
+            }
+        }
+    }
+
+    /// Updates or inserts Readonly `SystemPlugin` entities in specified storage.
+    /// Also removes stale plugins that no longer exist in remote plugin list.
+    ///
+    func upsertSystemPlugins(siteID: Int64, readonlySystemPlugins: [SystemPlugin], in storage: StorageType) {
+        readonlySystemPlugins.forEach { readonlySystemPlugin in
+            // load or create new StorageSystemPlugin matching the readonly one
+            let storageSystemPlugin: StorageSystemPlugin = {
+                if let systemPlugin = storage.loadSystemPlugin(siteID: readonlySystemPlugin.siteID, name: readonlySystemPlugin.name) {
+                    return systemPlugin
+                }
+                return storage.insertNewObject(ofType: StorageSystemPlugin.self)
+            }()
+
+            storageSystemPlugin.update(with: readonlySystemPlugin)
+        }
+
+        // remove stale system plugins
+        let currentSystemPlugins = readonlySystemPlugins.map(\.name)
+        storage.deleteStaleSystemPlugins(siteID: siteID, currentSystemPlugins: currentSystemPlugins)
+    }
+
+    /// Retrieve `SystemPlugin` entities of a specified storage by siteID
+    ///
+    func fetchSystemPlugins(siteID: Int64, completionHandler: @escaping ([SystemPlugin]?) -> Void) {
+        let writerStorage = storageManager.writerDerivedStorage
+        let systemPlugins = writerStorage.loadSystemPlugins(siteID: siteID).map { $0.toReadOnly() }
```

Comment:
⚠️ We should not be using a writer-derived storage to read data. It's probably better to use `storageManager.viewStorage`